trieste/acquisition/optimizer.py

```diff
@@ -192,73 +194,182 @@ def optimize_continuous(
+    batch calculations with TensorFlow, this function uses the Greenlet package
+    to run each individual optimization on micro-threads.
+
+    L-BFGS-B updates for each individual optimization are performed by
+    independent greenlets working with Numpy arrays, however, the evaluation
+    of our acquisition function (and its gradients) is calculated in parallel
+    (for each optimization step) using Tensorflow.
+
+    For :class:'TaggedProductSearchSpace' we only apply gradient updates to
+    its class:'Box' subspaces, fixing the discrete elements to the best values
+    found across the initial random search. To fix these discrete elements, we
+    optimize over a continuous class:'Box' relaxation of the discrete subspaces
+    which has equal upper and lower bounds, i.e. we specify an equality constraint
+    for this dimension in the scipy optimizer.
+
+    :param target_func: The function to maximise, with input shape [..., 1, D] and
+        output shape [..., 1].
+    :param space: The original search space.
+    :param starting_points: The points at which to begin our optimizations. The
+        leading dimension of `starting_points` controls the number of individual
+        optimization runs.
+    :param optimizer_args: Keyword arguments to pass to the Scipy optimizer.
+    :return: A list containing a Scipy OptimizeResult object for each optimization.
+    """
+
+    tf_dtype = starting_points.dtype  # type for communication with Trieste
+
+    num_optimization_runs = tf.shape(starting_points)[0].numpy()
+
+    def _objective_value(x: TensorType) -> TensorType:
+        return -target_func(tf.expand_dims(x, 1))  # [len(x),1]
+
+    def _objective_value_and_gradient(x: TensorType) -> Tuple[TensorType, TensorType]:
+        return tfp.math.value_and_gradient(_objective_value, x)  # [len(x), 1], [len(x), D]
+
+    if isinstance(
+        space, TaggedProductSearchSpace
+    ):  # build continuous relaxation of discrete subspaces
+        bounds = [
+            get_bounds_of_box_relaxation_around_point(space, starting_points[i : i + 1])
+            for i in tf.range(num_optimization_runs)
+        ]
+    else:
+        bounds = [spo.Bounds(space.lower, space.upper)] * num_optimization_runs
+
+    # Initialize the numpy arrays to be passed to the greenlets
+    np_batch_x = np.zeros((num_optimization_runs, tf.shape(starting_points)[1]), dtype=np.float64)
+    np_batch_y = np.zeros((num_optimization_runs,), dtype=np.float64)
+    np_batch_dy_dx = np.zeros(
+        (num_optimization_runs, tf.shape(starting_points)[1]), dtype=np.float64
+    )
+
+    # Set up child greenlets
+    child_greenlets = [ScipyLbfgsBGreenlet() for _ in range(num_optimization_runs)]
+    child_results: List[Union[spo.OptimizeResult, np.ndarray]] = [
+        gr.switch(starting_points[i].numpy(), bounds[i], optimizer_args)
+        for i, gr in enumerate(child_greenlets)
+    ]
+
+    while True:
```

Comment:
Is it guaranteed that this loop will always stop? It isn't immediately obvious, or I am just slow

Reply:
yep. eventually all the greenlets will do their maximum allowed number of steps